polkadot/runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -686,6 +617,110 @@ impl<T: Config> Pallet<T> {
 		};
 		Ok((processed, Some(all_weight_after).into()))
 	}
+
+	fn back_candidates(
+		context: ProcessInherentDataContext,
+		concluded_invalid_hashes: BTreeSet<CandidateHash>,
+		backed_candidates: Vec<BackedCandidate<T::Hash>>,
+	) -> Result<
+		(
+			Vec<(CandidateReceipt<T::Hash>, Vec<(ValidatorIndex, ValidityAttestation)>)>,
+			BTreeMap<ParaId, Vec<(BackedCandidate<T::Hash>, CoreIndex)>>,
+		),
+		DispatchErrorWithPostInfo,
+	> {
+		let allowed_relay_parents = shared::AllowedRelayParents::<T>::get();
+		let upcoming_new_session = initializer::Pallet::<T>::upcoming_session_change();
+
+		METRICS.on_candidates_processed_total(backed_candidates.len() as u64);
+
+		let initial_candidate_count = backed_candidates.len();
+
+		if !upcoming_new_session {
+			let occupied_cores =
+				inclusion::Pallet::<T>::get_occupied_cores().map(|(core, _para)| core).collect();
+
+			let mut eligible: BTreeMap<ParaId, BTreeSet<CoreIndex>> = BTreeMap::new();
+			let mut total_eligible_cores = 0;
+
+			for (core_idx, para_id) in scheduler::Pallet::<T>::eligible_paras(&occupied_cores) {
+				total_eligible_cores += 1;
+				log::trace!(target: LOG_TARGET, "Found eligible para {:?} on core {:?}", para_id, core_idx);
+				eligible.entry(para_id).or_default().insert(core_idx);
+			}
+
+			let node_features = configuration::ActiveConfig::<T>::get().node_features;
+			let core_index_enabled = node_features
+				.get(FeatureIndex::ElasticScalingMVP as usize)
+				.map(|b| *b)
+				.unwrap_or(false);
+
+			let allow_v2_receipts = node_features
+				.get(FeatureIndex::CandidateReceiptV2 as usize)
+				.map(|b| *b)
+				.unwrap_or(false);
+
+			let backed_candidates_with_core = sanitize_backed_candidates::<T>(
+				backed_candidates,
+				&allowed_relay_parents,
+				concluded_invalid_hashes,
+				eligible,
+				core_index_enabled,
+				allow_v2_receipts,
+			);
+			let count = count_backed_candidates(&backed_candidates_with_core);
+
+			ensure!(count <= total_eligible_cores, Error::<T>::UnscheduledCandidate);
+
+			METRICS.on_candidates_sanitized(count as u64);
+
+			// In `Enter` context (invoked during execution) no more candidates should be
+			// filtered, because they have already been filtered during `ProvideInherent`
+			// context. Abort in such cases.
+			if context == ProcessInherentDataContext::Enter {
```

Comment:
These checks should be top-level. They should be a wrapper around `process_inherent`. For debugging we can leave logs on when something was filtered for what reason.

Reply:
Reasoning is simple: If this is handled once at the top-level - you can not mess it up/forget it in some place.

Reply:
done. have a look